File: integration/bindings/client-js/actions.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import {
  ActionRegistry,
  ActionError,
} from "../../../packages/client-js/actions.mts";
import { Client } from "../../../packages/client-js/index.mts";
import { mkdtemp, rm, readFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { createRequire } from "node:module";
import { createClient } from "../../../packages/client-js/runtime.mts";
import { Transaction } from "../../../packages/client-js/transaction.mts";

test("observer retains a live wait, settles once, and caches the outcome", async () => {
  const registry = new ActionRegistry();
  const call = registry.register("one", (value) => ({ title: value.title }));
  assert.equal(call.status, "pending");
  const pending = call.wait();
  registry.complete({
    callId: "one",
    outcome: { status: "succeeded", result: { title: "A" } },
  });
  const first = await pending;
  assert.deepEqual(first, { result: { title: "A" }, error: null });
  assert.equal(call.status, "succeeded");
  assert.strictEqual(await call.wait(), first);
  assert.equal(registry.activeCount, 0);
});

test("observer converts backend failures and decoder failures into terminal outcomes", async () => {
  const registry = new ActionRegistry();
  const rejected = registry.register("rejected", (value) => value);
  const malformed = registry.register("malformed", () => {
    throw Error("bad result");
  });
  registry.complete({
    callId: "rejected",
    outcome: { status: "failed", code: "denied", execution: "rejected" },
  });
  registry.complete({
    callId: "malformed",
    outcome: { status: "succeeded", result: {} },
  });
  assert.equal((await rejected.wait()).error.code, "denied");
  assert.equal(
    (await malformed.wait()).error.code,
    "action.observation_failed",
  );
  assert.equal(malformed.status, "failed");
});

test("decoded results remain snapshots after the source object changes", async () => {
  const registry = new ActionRegistry();
  const call = registry.register("snapshot", (value) => ({
    title: value.title,
  }));
  const source = { title: "A" };
  registry.complete({
    callId: "snapshot",
    outcome: { status: "succeeded", result: source },
  });
  source.title = "B";
  assert.equal((await call.wait()).result.title, "A");
});

test("closing settles held handles even before wait is called", async () => {
  const registry = new ActionRegistry();
  const call = registry.register("one", (value) => value);
  registry.close();
  assert.equal(call.status, "failed");
  assert.equal((await call.wait()).error.code, "client.closed");
});

test("weak routing sweeps dead states but active waits remain live", async () => {
  const references = [];
  const registry = new ActionRegistry((state) => {
    const reference = {
      state,
      deref() {
        return this.state;
      },
    };
    references.push(reference);
    return reference;
  });
  registry.register("abandoned", (value) => value);
  references[0].state = undefined;
  registry.register("other", (value) => value);
  assert.equal(registry.routingCount, 1);
  let live = registry.register("live", (value) => value);
  const pending = live.wait();
  live = null;
  references[2].state = undefined;
  registry.complete({
    callId: "live",
    outcome: { status: "succeeded", result: 3 },
  });
  assert.equal((await pending).result, 3);
  assert.equal(registry.activeCount, 0);
});

test("missing WeakRef rejects before registration", () => {
  const registry = new ActionRegistry(null);
  assert.throws(
    () => registry.assertSupported(),
    (error) =>
      error instanceof ActionError &&
      error.code === "action.unsupported_runtime",
  );
  assert.equal(registry.routingCount, 0);
});

async function withClient(body) {
  const directory = await mkdtemp(join(tmpdir(), "axton-action-observer-"));
  const schema = JSON.parse(
    await readFile(
      new URL("../../../fixtures/schemas/entry.json", import.meta.url),
      "utf8",
    ),
  );
  schema.actions = [{ name: "Ping", version: 1, inputs: [], outputs: [] }];
  const client = await Client.open({ path: join(directory, "db"), schema });
  try {
    await body(client);
  } finally {
    await client.close();
    await rm(directory, { recursive: true, force: true });
  }
}

test("durable invocation registers before wake and drop settles its handle", async () => {
  await withClient(async (client) => {
    const call = await client.invokeAction("Ping", 1, {}, () => undefined);
    assert.equal(call.status, "pending");
    const waiting = call.wait();
    const tasks = await client.syncState();
    assert.equal(tasks.pending, 1);
    await client.drop((await client.pendingTasks())[0]?.ordinal ?? 1);
    const outcome = await waiting;
    assert.equal(outcome.error.code, "dropped");
    assert.equal(call.status, "failed");
    assert.equal((await client.syncState()).pending, 0);
  });
});

test("close fails a durable handle before wait is called", async () => {
  await withClient(async (client) => {
    const call = await client.invokeAction("Ping", 1, {}, () => undefined);
    await client.close();
    assert.equal((await call.wait()).error.code, "client.closed");
  });
});

test("standalone local writes stay out of the durable queue and reach reads and watch", async () => {
  await withClient(async (client) => {
    const observed = [];
    const stop = client.watch("Entry", {}, (rows) => observed.push(rows));
    try {
      await client.direct({
        model: "Entry",
        op: "create",
        identity: { id: "one" },
        values: { text: "A" },
      });
      assert.equal((await client.read("Entry", { id: "one" })).text, "A");
      assert.equal((await client.syncState()).pending, 0);
      await new Promise((resolve) => setImmediate(resolve));
      assert.ok(observed.some((rows) => rows.some((row) => row.text === "A")));
    } finally {
      stop();
    }
  });
});

test("the real native pump settles a waiting handle before a throwing diagnostic listener", async () => {
  const native = createRequire(import.meta.url)(
    "../../../bindings/node/axton-node.node",
  );
  const directory = await mkdtemp(join(tmpdir(), "axton-action-pump-"));
  const schema = {
    enums: [],
    models: [],
    actions: [{ name: "Ping", version: 1, inputs: [], outputs: [] }],
  };
  let requests = 0;
  const PumpClient = createClient(native, Transaction, () => ({
    open() {},
    push: async (kind, bodyText) => {
      const body = JSON.parse(bodyText);
      if (kind !== "push") throw Error(`unexpected ${kind}`);
      requests++;
      return JSON.stringify({
        clientId: body.clientId,
        batchSequence: body.batchSequence,
        rejections: [],
        records: [],
        completions: body.mutations.map((mutation) => ({
          callId: mutation.callId,
          outcome: { status: "succeeded", result: null },
        })),
      });
    },
  }));
  const client = await PumpClient.open({ path: join(directory, "db"), schema });
  try {
    const connection = await client.connect({
      url: "http://unused",
      token: "token",
    });
    client.onActionCompletion(() => {
      throw Error("diagnostic failed");
    });
    const call = await client.invokeAction("Ping", 1, {}, () => undefined);
    const outcome = await Promise.race([
      call.wait(),
      new Promise((_, reject) =>
        setTimeout(() => reject(Error("wait timed out")), 1000),
      ),
    ]);
    assert.deepEqual(outcome, { result: undefined, error: null });
    assert.equal(call.status, "succeeded");
    assert.equal(requests, 1);
    assert.equal((await client.syncState()).pending, 0);
    await connection.close();
  } finally {
    await client.close();
    await rm(directory, { recursive: true, force: true });
  }
});

test("direct invocation decodes a committed response and maps unavailable transport errors", async () => {
  const native = createRequire(import.meta.url)(
    "../../../bindings/node/axton-node.node",
  );
  const directory = await mkdtemp(join(tmpdir(), "axton-action-direct-"));
  const schema = {
    enums: [],
    models: [],
    actions: [{ name: "Ping", version: 1, inputs: [], outputs: [] }],
  };
  const DirectClient = createClient(native, Transaction, () => ({
    open() {},
    push: async (kind, bodyText) => {
      if (kind !== "action") throw Error(`unexpected ${kind}`);
      const body = JSON.parse(bodyText);
      return JSON.stringify({
        completion: {
          callId: body.call.callId,
          outcome: { status: "succeeded", result: null },
        },
        records: [],
      });
    },
  }));
  const client = await DirectClient.open({
    path: join(directory, "db"),
    schema,
  });
  try {
    await assert.rejects(
      client.invokeDirectAction("Ping", 1, {}, () => undefined),
      (error) =>
        error instanceof ActionError && error.code === "action.unavailable",
    );
    const connection = await client.connect({
      url: "http://unused",
      token: "token",
    });
    assert.equal(
      await client.invokeDirectAction("Ping", 1, {}, () => undefined),
      undefined,
    );
    assert.equal((await client.syncState()).pending, 0);
    await connection.close();
  } finally {
    await client.close();
    await rm(directory, { recursive: true, force: true });
  }
});

test("unsupported runtime fails before durable submission", async () => {
  await withClient(async (client) => {
    const original = globalThis.WeakRef;
    try {
      globalThis.WeakRef = undefined;
      await assert.rejects(
        client.invokeAction("Ping", 1, {}, () => undefined),
        (error) => error.code === "action.unsupported_runtime",
      );
    } finally {
      globalThis.WeakRef = original;
    }
    assert.equal((await client.syncState()).pending, 0);
  });
});

test("standalone direct writes and Actions reject from an active transaction callback", async () => {
  await withClient(async (client) => {
    await client.transaction(async () => {
      await assert.rejects(
        client.direct({
          model: "Entry",
          op: "create",
          identity: { id: "blocked" },
          values: { text: "A" },
        }),
        /transaction_active/,
      );
      await assert.rejects(
        client.invokeAction("Ping", 1, {}, () => undefined),
        (error) =>
          error instanceof ActionError && error.code === "transaction_active",
      );
      await assert.rejects(
        client.invokeDirectAction("Ping", 1, {}, () => undefined),
        (error) =>
          error instanceof ActionError && error.code === "transaction_active",
      );
    });
    assert.equal((await client.syncState()).pending, 0);
  });
});

test("close racing a committed submit still yields a terminal handle", async () => {
  const binding = createRequire(import.meta.url)(
    "../../../bindings/node/axton-node.node",
  );
  const directory = await mkdtemp(join(tmpdir(), "axton-action-close-race-"));
  const schema = {
    enums: [],
    models: [],
    actions: [{ name: "Ping", version: 1, inputs: [], outputs: [] }],
  };
  let entered;
  let release;
  const submitted = new Promise((resolve) => {
    entered = resolve;
  });
  const gate = new Promise((resolve) => {
    release = resolve;
  });
  const native = {
    async clientCall(request) {
      const result = await binding.clientCall(request);
      if (JSON.parse(request).op === "submitAction") {
        entered();
        await gate;
      }
      return result;
    },
  };
  const RaceClient = createClient(native, Transaction, () => {
    throw Error("network not configured");
  });
  const client = await RaceClient.open({ path: join(directory, "db"), schema });
  try {
    const callPromise = client.invokeAction("Ping", 1, {}, () => undefined);
    await submitted;
    const closing = client.close();
    release();
    const call = await callPromise;
    assert.equal((await call.wait()).error.code, "client.closed");
    await closing;
  } finally {
    release();
    await client.close();
    await rm(directory, { recursive: true, force: true });
  }
});

File: packages/dart/test/actions_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:axton/axton.dart';
import 'package:axton/src/actions.dart' show ActionObservers, ActionWeakState;
import 'package:test/test.dart';

void main() {
  test('weak routes sweep without retaining an abandoned handle', () {
    final refs = <_TestWeak>[];
    final observers = ActionObservers(
      weak: (state) {
        final ref = _TestWeak(state);
        refs.add(ref);
        return ref;
      },
    );
    observers.register<void>('gone', (_) {});
    refs.single.value = null;
    expect(observers.routingCount, 0);
    observers.complete({
      'callId': 'gone',
      'outcome': {'status': 'succeeded', 'result': null},
    });
  });

  test('active wait survives loss of its weak routing target', () async {
    final refs = <_TestWeak>[];
    final observers = ActionObservers(
      weak: (state) {
        final ref = _TestWeak(state);
        refs.add(ref);
        return ref;
      },
    );
    final call = observers.register<String>('held', (value) => value as String);
    final waiting = call.wait();
    refs.single.value = null;
    observers.complete({
      'callId': 'held',
      'outcome': {'status': 'succeeded', 'result': 'ready'},
    });
    final result =
        await waiting.timeout(const Duration(milliseconds: 100))
            as ActionSuccess<String>;
    expect(result.result, 'ready');
  });

  late Directory directory;
  late Client client;
  setUp(() async {
    directory = await Directory.systemTemp.createTemp('axton-actions-');
    client = await Client.open(
      path: '${directory.path}/db',
      schema: {
        'enums': [],
        'models': [],
        'actions': [
          {'name': 'Ping', 'version': 1, 'inputs': [], 'outputs': []},
        ],
      },
      libraryPath: Platform.environment['AXTON_LIBRARY']!,
    );
  });
  tearDown(() async {
    await client.close();
    await directory.delete(recursive: true);
  });

  test('real native acknowledgement settles a cached typed handle', () async {
    final call = await client.invokeAction<void>('Ping', 1, {}, (_) {});
    expect(call.status, ActionStatus.pending);
    final frozen = await client.freeze();
    expect(frozen, isNotNull);
    final request = jsonDecode(frozen!) as Map<String, dynamic>;
    final mutation = (request['mutations'] as List).single as Map;
    final receipt = {
      'clientId': request['clientId'],
      'batchSequence': request['batchSequence'],
      'rejections': [],
      'completions': [
        {
          'callId': mutation['callId'],
          'outcome': {'status': 'succeeded', 'result': null},
        },
      ],
      'records': [],
    };
    final waiting = call.wait();
    await client.acknowledge(request['batchSequence'] as int, receipt);
    expect(await waiting, isA<ActionSuccess<void>>());
    expect(identical(await waiting, await call.wait()), isTrue);
    expect(call.status, ActionStatus.succeeded);
  });

  test('store selector travels beside args on both routes', () async {
    await expectLater(
      client.invokeAction<void>(
        'Ping',
        1,
        {},
        (_) {},
        store: const _Store({'missing': false}),
      ),
      throwsA(isA<ActionError>()),
    );
    expect(await client.freeze(), isNull, reason: 'nothing was enqueued');
    await client.invokeAction<void>(
      'Ping',
      1,
      {},
      (_) {},
      store: const _Store(false),
    );
    await client.invokeAction<void>(
      'Ping',
      1,
      {},
      (_) {},
      store: const _Store(null),
    );
    final frozen = jsonDecode((await client.freeze())!) as Map;
    final mutations = (frozen['mutations'] as List).cast<Map>();
    expect(mutations[0]['store'], false);
    expect(mutations[0]['args'], isEmpty);
    expect(mutations[1].containsKey('store'), isFalse);
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    final bodies = <Map>[];
    final served = server.listen((request) async {
      if (request.uri.path != '/sync/actions') {
        request.response.statusCode = 404;
        await request.response.close();
        return;
      }
      final body = jsonDecode(await utf8.decoder.bind(request).join()) as Map;
      bodies.add(body);
      request.response.write(
        jsonEncode({
          'completion': {
            'callId': (body['call'] as Map)['callId'],
            'outcome': {'status': 'succeeded', 'result': null},
          },
          'records': [],
        }),
      );
      await request.response.close();
    });
    try {
      final connection = await client.connect(
        SyncServer(url: 'http://127.0.0.1:${server.port}', token: () => 'a'),
      );
      try {
        await client.invokeDirectAction<void>(
          'Ping',
          1,
          {},
          (_) {},
          store: const _Store(false),
        );
      } finally {
        await connection.close();
      }
      final direct = bodies.single;
      expect((direct['call'] as Map)['store'], false);
      expect((direct['call'] as Map)['args'], isEmpty);
    } finally {
      await served.cancel();
      await server.close(force: true);
    }
  });

  test('close settles a handle even without a prior wait', () async {
    final call = await client.invokeAction<void>('Ping', 1, {}, (_) {});
    await client.close();
    final outcome = await call.wait();
    expect(outcome, isA<ActionFailure<void>>());
    expect((outcome as ActionFailure<void>).error.code, 'client.closed');
    expect(call.status, ActionStatus.failed);
  });

  test('close racing a queued submit returns a failed handle', () async {
    final entered = Completer<void>();
    final release = Completer<void>();
    final blocker = client.transaction((_) async {
      entered.complete();
      await release.future;
    });
    await entered.future;
    final submitting = client.invokeAction<void>('Ping', 1, {}, (_) {});
    final closing = client.close();
    release.complete();
    await blocker;
    final call = await submitting.timeout(const Duration(seconds: 1));
    final failure = await call.wait() as ActionFailure<void>;
    expect(failure.error.code, 'client.closed');
    await closing;
  });

  test('drop settles a pending handle through native completion', () async {
    final call = await client.invokeAction<void>('Ping', 1, {}, (_) {});
    await client.drop(1);
    final outcome = await call.wait() as ActionFailure<void>;
    expect(outcome.error.code, 'dropped');
  });

  test('rebuild discard settles live unsent and frozen handles', () async {
    final schema =
        jsonDecode(
              await File('../../fixtures/schemas/entry.json').readAsString(),
            )
            as Map<String, dynamic>;
    schema['actions'] = [
      {'name': 'Ping', 'version': 1, 'inputs': [], 'outputs': []},
    ];
    final breaking = jsonDecode(jsonEncode(schema)) as Map<String, dynamic>;
    ((breaking['models'] as List).first['fields'] as List).add({
      'name': 'due',
      'nullable': false,
      'type': {'kind': 'scalar', 'name': 'string'},
    });
    for (final frozen in [false, true]) {
      final path = '${directory.path}/${frozen ? 'frozen' : 'unsent'}';
      final old = await Client.open(
        path: path,
        schema: schema,
        libraryPath: Platform.environment['AXTON_LIBRARY']!,
      );
      await old.invokeAction<void>('Ping', 1, {}, (_) {});
      await old.close();
      final reopened = await Client.open(
        path: path,
        schema: breaking,
        libraryPath: Platform.environment['AXTON_LIBRARY']!,
      );
      try {
        final call = await reopened.invokeAction<void>('Ping', 1, {}, (_) {});
        final waiting = call.wait();
        if (frozen) await reopened.freeze();
        final report = await reopened.rebuild(discardPending: true);
        expect(
          (report['abandonedCalls'] as List).any(
            (entry) => entry['frozen'] == frozen,
          ),
          isTrue,
        );
        final outcome = await waiting as ActionFailure<void>;
        expect(outcome.error.code, 'abandoned');
        expect(outcome.error.execution, frozen ? 'unknown' : 'rejected');
        expect(call.status, ActionStatus.failed);
      } finally {
        await reopened.close();
      }
    }
  });

  test(
    'dropping an unsent Action settles its live lifecycle dependent',
    () async {
      final schema =
          jsonDecode(
                await File('../../fixtures/schemas/entry.json').readAsString(),
              )
              as Map<String, dynamic>;
      schema['actions'] = [
        {
          'name': 'Add',
          'version': 1,
          'inputs': [
            {
              'kind': 'model',
              'name': 'entry',
              'model': 'Entry',
              'operation': 'create',
              'cardinality': 'single',
            },
          ],
          'outputs': [],
        },
        {
          'name': 'Edit',
          'version': 1,
          'inputs': [
            {
              'kind': 'model',
              'name': 'entry',
              'model': 'Entry',
              'operation': 'update',
              'cardinality': 'single',
            },
          ],
          'outputs': [],
        },
      ];
      final local = await Client.open(
        path: '${directory.path}/dependent',
        schema: schema,
        libraryPath: Platform.environment['AXTON_LIBRARY']!,
      );
      try {
        final created = await local.invokeAction<void>('Add', 1, {
          'entry': {'id': 'e', 'text': 'A'},
        }, (_) {});
        final dependent = await local.invokeAction<void>('Edit', 1, {
          'entry': {'id': 'e', 'text': 'B'},
        }, (_) {});
        final waiting = dependent.wait();
        final state = await local.recordSyncState('Entry', {'id': 'e'});
        await local.drop((state['pending'] as List).first['ordinal'] as int);
        expect(
          (await created.wait() as ActionFailure<void>).error.code,
          'dropped',
        );
        expect(
          (await waiting as ActionFailure<void>).error.code,
          'dependency.rejected',
        );
        expect(dependent.status, ActionStatus.failed);
        expect((await local.syncState())['pending'], 0);
      } finally {
        await local.close();
      }
    },
  );

  test(
    'typed actions and standalone direct reject inside a transaction',
    () async {
      await client.transaction((tx) async {
        for (final pending in <Future<dynamic>>[
          client.invokeAction<void>('Ping', 1, {}, (_) {}),
          client.invokeDirectAction<void>('Ping', 1, {}, (_) {}),
        ]) {
          await expectLater(
            pending.timeout(const Duration(milliseconds: 300)),
            throwsA(
              isA<ActionError>()
                  .having((e) => e.code, 'code', 'transaction_active')
                  .having((e) => e.execution, 'execution', 'rejected'),
            ),
          );
        }
        await expectLater(
          client
              .direct({
                'model': 'Entry',
                'op': 'delete',
                'identity': {'id': 'e'},
              })
              .timeout(const Duration(milliseconds: 300)),
          throwsA(
            isA<StateError>().having(
              (e) => e.message,
              'message',
              'transaction_active',
            ),
          ),
        );
      });
    },
  );

  test(
    'invalid result decoder settles the handle as observation failure',
    () async {
      final call = await client.invokeAction<String>('Ping', 1, {}, (_) {
        throw const FormatException('bad result');
      });
      final request = jsonDecode((await client.freeze())!) as Map;
      final mutation = (request['mutations'] as List).single as Map;
      await client.acknowledge(request['batchSequence'] as int, {
        'clientId': request['clientId'],
        'batchSequence': request['batchSequence'],
        'rejections': [],
        'completions': [
          {
            'callId': mutation['callId'],
            'outcome': {'status': 'succeeded', 'result': null},
          },
        ],
        'records': [],
      });
      final outcome = await call.wait() as ActionFailure<String>;
      expect(outcome.error.code, 'action.observation_failed');
      expect(outcome.error.cause, isA<FormatException>());
      expect(call.status, ActionStatus.failed);
    },
  );

  test(
    'direct action returns decoded result and maps unavailable transport',
    () async {
      await expectLater(
        client.invokeDirectAction<void>('Ping', 1, {}, (_) {}),
        throwsA(
          isA<ActionError>().having(
            (e) => e.code,
            'code',
            'action.unavailable',
          ),
        ),
      );
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      var succeed = true;
      final sub = server.listen((request) async {
        if (request.uri.path != '/sync/actions') {
          request.response.statusCode = 404;
          await request.response.close();
          return;
        }
        final body = jsonDecode(await utf8.decoder.bind(request).join()) as Map;
        request.response.write(
          jsonEncode({
            'completion': {
              'callId': (body['call'] as Map)['callId'],
              'outcome': succeed
                  ? {'status': 'succeeded', 'result': null}
                  : {
                      'status': 'failed',
                      'code': 'handler.failed',
                      'execution': 'rejected',
                    },
            },
            'records': [],
          }),
        );
        await request.response.close();
      });
      final connection = await client.connect(
        SyncServer(
          url: 'http://127.0.0.1:${server.port}',
          token: () => 'alice',
        ),
      );
      try {
        expect(
          await client.invokeDirectAction<String>(
            'Ping',
            1,
            {},
            (_) => 'decoded',
          ),
          'decoded',
        );
        await expectLater(
          client.invokeDirectAction<String>('Ping', 1, {}, (_) {
            throw const FormatException('bad direct result');
          }),
          throwsA(
            isA<ActionError>()
                .having((e) => e.code, 'code', 'action.observation_failed')
                .having((e) => e.cause, 'cause', isA<FormatException>()),
          ),
        );
        succeed = false;
        await expectLater(
          client.invokeDirectAction<void>('Ping', 1, {}, (_) {}),
          throwsA(
            isA<ActionError>()
                .having((e) => e.code, 'code', 'handler.failed')
                .having((e) => e.execution, 'execution', 'rejected'),
          ),
        );
      } finally {
        await connection.close();
        await sub.cancel();
        await server.close(force: true);
      }
    },
  );

  test('connected native push pump resolves a live typed handle', () async {
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    var executions = 0;
    final served = server.listen((request) async {
      if (request.uri.path != '/sync/mutations') {
        request.response.statusCode = 404;
        await request.response.close();
        return;
      }
      final body = jsonDecode(await utf8.decoder.bind(request).join()) as Map;
      final mutation = (body['mutations'] as List).single as Map;
      executions++;
      request.response.write(
        jsonEncode({
          'clientId': body['clientId'],
          'batchSequence': body['batchSequence'],
          'rejections': [],
          'completions': [
            {
              'callId': mutation['callId'],
              'outcome': {'status': 'succeeded', 'result': null},
            },
          ],
          'records': [],
        }),
      );
      await request.response.close();
    });
    final connection = await client.connect(
      SyncServer(url: 'http://127.0.0.1:${server.port}', token: () => 'alice'),
    );
    try {
      final call = await client.invokeAction<void>('Ping', 1, {}, (_) {});
      final deadline = DateTime.now().add(const Duration(seconds: 2));
      while (call.status == ActionStatus.pending &&
          DateTime.now().isBefore(deadline)) {
        await Future<void>.delayed(const Duration(milliseconds: 5));
      }
      expect(
        call.status,
        ActionStatus.succeeded,
        reason: 'the pump runs without an active wait',
      );
      expect(
        await call.wait().timeout(const Duration(seconds: 2)),
        isA<ActionSuccess<void>>(),
      );
      expect(executions, 1);
      expect((await client.syncState())['pending'], 0);
    } finally {
      await connection.close();
      await served.cancel();
      await server.close(force: true);
    }
  });

  test(
    'durable reports isolate throwing onError and keep the pump alive',
    () async {
      final schema =
          jsonDecode(
                await File('../../fixtures/schemas/entry.json').readAsString(),
              )
              as Map<String, dynamic>;
      schema['actions'] = [
        {'name': 'Ping', 'version': 1, 'inputs': [], 'outputs': []},
      ];
      final local = await Client.open(
        path: '${directory.path}/durable-diagnostic',
        schema: schema,
        libraryPath: Platform.environment['AXTON_LIBRARY']!,
      );
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      var requests = 0;
      final served = server.listen((request) async {
        if (request.uri.path != '/sync/mutations') {
          request.response.statusCode = 404;
          await request.response.close();
          return;
        }
        final body = jsonDecode(await utf8.decoder.bind(request).join()) as Map;
        requests++;
        request.response.write(
          jsonEncode({
            'clientId': body['clientId'],
            'batchSequence': body['batchSequence'],
            'rejections': [],
            'completions': [
              for (final mutation in body['mutations'] as List)
                {
                  'callId': (mutation as Map)['callId'],
                  'outcome': {'status': 'succeeded', 'result': null},
                },
            ],
            'records': requests == 3
                ? []
                : [
                    for (final id in ['a', 'b'])
                      {
                        'model': 'Entry',
                        'identity': {'id': id},
                        'stamp': 1,
                        'state': {
                          'text': requests == 1 ? 'first' : 'second',
                          'note': null,
                        },
                      },
                  ],
          }),
        );
        await request.response.close();
      });
      final diagnostic = StateError('application diagnostic failed');
      final reports = <Object>[];
      final observed = <Object>[];
      final finished = Completer<void>();
      try {
        runZonedGuarded(() {
          () async {
            final connection = await local.connect(
              SyncServer(
                url: 'http://127.0.0.1:${server.port}',
                token: () => 'alice',
              ),
              onError: (report) {
                reports.add(report);
                throw diagnostic;
              },
            );
            try {
              for (var i = 0; i < 3; i++) {
                final call = await local.invokeAction<void>(
                  'Ping',
                  1,
                  {},
                  (_) {},
                );
                expect(
                  await call.wait().timeout(const Duration(seconds: 2)),
                  isA<ActionSuccess<void>>(),
                );
              }
            } finally {
              await connection.close();
            }
          }().then(finished.complete, onError: finished.completeError);
        }, (error, stack) => observed.add(error));
        await finished.future.timeout(const Duration(seconds: 3));
        expect(requests, 3);
        expect(reports, hasLength(2));
        expect(
          reports.every((report) => report.toString().contains('conflict')),
          isTrue,
        );
        expect(observed, [same(diagnostic), same(diagnostic)]);
        expect((await local.syncState())['pending'], 0);
      } finally {
        await local.close();
        await served.cancel();
        await server.close(force: true);
      }
    },
  );

  test(
    'throwing direct diagnostic preserves applied result and reaches the Zone',
    () async {
      final schema =
          jsonDecode(
                await File('../../fixtures/schemas/entry.json').readAsString(),
              )
              as Map<String, dynamic>;
      schema['actions'] = [
        {'name': 'Ping', 'version': 1, 'inputs': [], 'outputs': []},
      ];
      final local = await Client.open(
        path: '${directory.path}/direct-diagnostic',
        schema: schema,
        libraryPath: Platform.environment['AXTON_LIBRARY']!,
      );
      final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      var requests = 0;
      final served = server.listen((request) async {
        if (request.uri.path != '/sync/actions') {
          request.response.statusCode = 404;
          await request.response.close();
          return;
        }
        final body = jsonDecode(await utf8.decoder.bind(request).join()) as Map;
        requests++;
        request.response.write(
          jsonEncode({
            'completion': {
              'callId': (body['call'] as Map)['callId'],
              'outcome': {'status': 'succeeded', 'result': null},
            },
            'records': [
              {
                'model': 'Entry',
                'identity': {'id': 'e'},
                'stamp': 1,
                'state': {
                  'text': requests == 1 ? 'first' : 'second',
                  'note': null,
                },
              },
            ],
          }),
        );
        await request.response.close();
      });
      final diagnostic = StateError('diagnostic failed');
      final observed = <Object>[];
      try {
        final connection = await local.connect(
          SyncServer(
            url: 'http://127.0.0.1:${server.port}',
            token: () => 'alice',
          ),
          onError: (_) => throw diagnostic,
        );
        try {
          expect(
            await local.invokeDirectAction<String>(
              'Ping',
              1,
              {},
              (_) => 'first',
            ),
            'first',
          );
          final result = Completer<String>();
          runZonedGuarded(() {
            local
                .invokeDirectAction<String>('Ping', 1, {}, (_) => 'decoded')
                .then(result.complete, onError: result.completeError);
          }, (error, stack) => observed.add(error));
          expect(
            await result.future.timeout(const Duration(seconds: 2)),
            'decoded',
          );
        } finally {
          await connection.close();
        }
        expect(observed, [same(diagnostic)]);
      } finally {
        await local.close();
        await served.cancel();
        await server.close(force: true);
      }
    },
  );

  test('standalone direct write commits locally and wakes watch', () async {
    final schema =
        jsonDecode(
              await File('../../fixtures/schemas/entry.json').readAsString(),
            )
            as Map<String, dynamic>;
    final local = await Client.open(
      path: '${directory.path}/local',
      schema: schema,
      libraryPath: Platform.environment['AXTON_LIBRARY']!,
    );
    try {
      final seen = <List<Map<String, dynamic>>>[];
      final first = Completer<void>();
      final second = Completer<void>();
      final sub = local.watch('Entry').listen((rows) {
        seen.add(rows);
        if (seen.length == 1) first.complete();
        if (seen.length == 2) second.complete();
      });
      await first.future;
      await local.direct({
        'model': 'Entry',
        'op': 'create',
        'identity': {'id': 'e'},
        'values': {'text': 'local'},
      });
      await second.future.timeout(const Duration(seconds: 1));
      expect((await local.read('Entry', {'id': 'e'}))?['text'], 'local');
      expect((seen.last.single)['text'], 'local');
      expect(await local.pendingTasks(), isEmpty);
      await sub.cancel();
    } finally {
      await local.close();
    }
  });
}

final class _TestWeak implements ActionWeakState {
  Object? value;
  _TestWeak(this.value);
  @override
  Object? get target => value;
}

final class _Store extends ActionStore {
  const _Store(this.wire);
  final Object? wire;
  @override
  Object? toWire() => wire;
}
